src/api/eth_subscribe.rs

```diff
@@ -0,0 +1,161 @@
+        .map(Async::Ready)
+        .map_err(Into::into),
+      Ok(Async::Ready(None)) => Ok(Async::Ready(None)),
+      Ok(Async::NotReady) => Ok(Async::NotReady),
+      Err(e) => Err(e),
+    }
+  }
+}
+
+impl<T: DuplexTransport, I> fmt::Debug for SubscriptionStream<T, I> {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!(f, "{:?}", self.id())
+  }
+}
+
+impl<T: DuplexTransport, I> Drop for SubscriptionStream<T, I> {
+  fn drop(&mut self) {
+    self.transport.unsubscribe(self.id());
+  }
+}
+
+#[derive(Debug)]
+pub struct SubscriptionResult<T: DuplexTransport, I> {
+  transport: T,
+  inner: CallResult<String, T::Out>,
+  _marker: PhantomData<I>,
+}
+
+impl<T: DuplexTransport, I> SubscriptionResult<T, I> {
+  pub fn new(transport: T, id_future: CallResult<String, T::Out>) -> Self {
+    SubscriptionResult {
+      transport,
+      inner: id_future,
+      _marker: PhantomData,
+    }
+  }
+}
+
+impl<T, I> Future for SubscriptionResult<T, I>
+where
+  T: DuplexTransport,
+  I: serde::de::DeserializeOwned,
+{
+  type Item = (String, SubscriptionStream<T, I>);
+  type Error = Error;
+
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    match self.inner.poll() {
+      Ok(Async::Ready(id)) => Ok(Async::Ready((
+        id.clone(),
+        SubscriptionStream::new(self.transport.clone(), id),
+      ))),
+      Ok(Async::NotReady) => Ok(Async::NotReady),
+      Err(e) => Err(e),
+    }
+  }
+}
+
+impl<T: DuplexTransport> EthSubscribe<T> {
+  /// Create a new heads subscription
+  pub fn subscribe_new_heads(&self) -> SubscriptionResult<T, BlockHeader> {
+    let subscription = helpers::serialize(&&"newHeads");
+    let id_future = CallResult::new(self.transport.execute("eth_subscribe", vec![subscription]));
+    SubscriptionResult::new(self.transport().clone(), id_future)
+  }
+
+  /// Create a logs subscription
+  pub fn subscribe_logs(
+    &self,
+    filter: Filter
+  ) -> SubscriptionResult<T, Log> {
+    let subscription = helpers::serialize(&&"logs");
+    let filter = helpers::serialize(&filter);
+    let id_future = CallResult::new(self.transport.execute("eth_subscribe", vec![subscription, filter]));
+    SubscriptionResult::new(self.transport().clone(), id_future)
+  }
+
+  pub fn subscribe_new_pending_transactions(&self) -> SubscriptionResult<T, H256> {
+    let subscription = helpers::serialize(&&"newPendingTransactions");
+    let id_future = CallResult::new(self.transport.execute("eth_subscribe", vec![subscription]));
+    SubscriptionResult::new(self.transport().clone(), id_future)
+  }
+
+  pub fn subscribe_syncing(&self) -> SubscriptionResult<T, rpc::Value> {
+    let subscription = helpers::serialize(&&"syncing");
+    let id_future = CallResult::new(self.transport.execute("eth_subscribe", vec![subscription]));
+    SubscriptionResult::new(self.transport().clone(), id_future)
+  }
+
+  pub fn unsubscribe(&self, id: &str) -> CallResult<bool, T::Out> {
```

Comment:
I would rather see people using `Stream::unsubscribe()` instead, hence imho better to hide the id